### ISSUE TITLE
Clippy should detect opportunities for `into_` variants

### ISSUE THREAD

**Issue by aeshirey** [rust-lang/rust-clippy#8130]
### What it does

Identify places where an `into_*` function variant might be more efficient than cases where a non-`into` plus a `clone` is being used.

### Lint Name

use_into_version

### Category

style, perf

### Advantage

- Removes unnecessary allocations of types by moving them
- Cleans up code

### Example

```rust
// Note: this function receives and returns owned values
fn get_keys(hm: HashMap<String, u32>) -> Vec<String> {
    hm.keys().map(|k| k.to_string()).collect()
}
```

Could be written as:

```rust
fn get_keys(hm: HashMap<String, u32>) -> Vec<String> {
    hm.into_keys().collect()
}
```

**Comment by aeshirey** [comment]
Not knowing the internals of clippy, this may be a tough nut to crack, as it would need to identify that an owned value that is being consumed isn't used subsequently used. For example:

```rust
fn maybe_sorted_keys(hm: HashMap<String, u32>, sort: bool) -> Vec<String> {
    // Assume the following line is the outcome of this proposed lint:
    let mut keys_only: Vec<_> = hm.into_keys().collect();
    keys_only.sort();

    if sort {
        keys_only
    } else {
        hm.keys().map(|k| k.clone()).collect()
//      ^^ value borrowed here after move
    }
}
```
